File: bench/luajit-browser/latency.mjs
import {startGuest} from './guest-runtime.mjs';
const output = document.querySelector('#result');
const terminal = document.querySelector('#terminal');
const query = new URL(location.href).searchParams;
const backend = query.get('backend') || 'luajit';
const trials = Number(query.get('trials') || 30);
const samples = [];
const phaseMarkers = [];
const schedule = [];
for (const kind of ['noop', 'check', 'compile', 'hover', 'edit', 'large-check', 'import-check']) {
  for (const count of (kind === 'large-check' || kind === 'import-check' ? [1, 4] : [1, 16])) {
    for (const bytes of (kind === 'noop' ? [0, 4096, 65536] : [0])) {
      for (let trial = -2; trial < trials; trial++) schedule.push({kind, count, bytes, trial});
    }
  }
}
const quantile = (values, p) => [...values].sort((a,b) => a-b)[Math.ceil(values.length*p)-1];
let sentAt, previous, index = 0, startup;
try {
  const handle = ({result}) => {
      if (!startup) startup = result;
      if (previous) {
        const elapsedMs = performance.now() - sentAt;
        if (result.count !== previous.count) throw new Error('Missing batch results');
        for (const answer of Object.values(result.answers)) {
          if (answer.reason || Object.values(answer.diagnostics || {}).some(d => d.severity === 'error')) {
            throw new Error('Compiler returned a diagnostic: ' + JSON.stringify(answer));
          }
        }
        if (previous.trial >= 0) samples.push({...previous, elapsedMs,
          sampledWallMs: result.sampledWallMs, guestCpuMs: result.guestCpuMs});
      }
      previous = schedule[index++];
      if (!previous) return {stop: true};
      terminal.textContent = JSON.stringify(previous);
      sentAt = performance.now();
      const source = previous.kind === 'large-check' ? 'local values={' + Array.from({length:2048}, (_,i)=>i+1).join(',') + '}; return values[1]'
        : previous.kind === 'import-check' ? 'local text = require("nupp.text"); return text' : undefined;
      return {...previous, source, kind: source ? 'check' : previous.kind, padding: 'x'.repeat(previous.bytes)};
  };
  if (backend === 'lua51') {
    const worker = new Worker('./latency-portable.mjs', {type:'module'});
    try {
      await new Promise((resolve,reject) => {
        worker.onerror = event => reject(new Error(event.message));
        worker.onmessage = ({data}) => {
          if (data.done) resolve();
          else if (data.error) reject(new Error(data.error));
          else { try { worker.postMessage(handle(data)); } catch (error) { reject(error); } }
        };
      });
    } finally { worker.terminate(); }
  } else {
    await startGuest({appUrl:'./latency.lua', config:{mode:'compiler',jit:false,slim:true,nativeCompiler:query.has('native'), direct:query.has('direct'), compilerTransport:query.has('transport'), timingProbe:query.has('phases'),
        bundle:query.has('native') ? '/nupp/native-compiler.ljbc' : undefined},
      deadlineMs:240000, effectHandlers:{'compiler-request':handle},
      onProgress:message => {if(message.log) terminal.textContent=message.log; if(message.type === 'phase') phaseMarkers.push(message);},
    }).result;
  }
  const groups = new Map();
  for (const sample of samples) {
    const key = `${sample.kind}/${sample.count}/${sample.bytes}`;
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(sample);
  }
  const summary = [...groups].map(([key, values]) => ({key,
    p50Ms: quantile(values.map(s => s.elapsedMs), .5), p95Ms: quantile(values.map(s => s.elapsedMs), .95),
    msPerOperation: values.reduce((sum,s) => sum+s.elapsedMs,0) / values.reduce((sum,s) => sum+s.count,0),
    zeroGuestTimerSamples: values.filter(s => s.sampledWallMs === 0).length,
    sampledWallP50Ms: quantile(values.map(s => s.sampledWallMs), .5),
  }));
  output.textContent = JSON.stringify({ok:true, backend, direct:query.has('direct'), compilerTransport:query.has('transport'), startup, summary, samples, phaseMarkers});
  output.dataset.status = 'passed';
} catch (error) {
  output.textContent = JSON.stringify({ok:false, error:String(error.stack || error), samples});
  output.dataset.status = 'failed';
}

File: bench/luajit-browser/latency.lua
-- A retained compiler session driven through the guest's actual effect transport.
local encode = dofile("/nupp/json-encoder.lua")()
local decode = dofile("/nupp/json-decoder.lua")()
local now = assert(__qemuNow)
local config = assert(__qemuConfig)
if not config.nativeCompiler then
    package.loaded["nupp.runtime.bitops"] = require("bit")
end
local sequence = 0

local function phase(name)
    if config.timingProbe then
        io.stdout:write("@@NUPP_PHASE@@ " .. sequence .. " " .. name .. "\n")
        io.stdout:flush()
    end
end

local before = now()
local Browser = assert(loadfile(config.bundle or "/nupp/playground-compiler.ljbc"))()
local session = Browser.new()
local result = {loadedMs = now() - before}
while true do
    phase("encode-start")
    local request
    if config.compilerTransport then
        assert(__qemuTransport).write("/host/request.json", encode(result))
        phase("encode-end")
        io.stdout:write("\n@@NUPP_COMPILER_FRAME@@ " .. sequence .. "\n")
        io.stdout:flush()
        assert(tonumber(io.read("*l")) == sequence, "stale compiler reply")
        phase("decode-start")
        request = decode(__qemuTransport.read("/host/response.json"))
        if request.payloadField then
            assert(request.payloadField == "source" or request.payloadField == "padding")
            request[request.payloadField] = __qemuTransport.read("/host/writeback.bin")
        end
        sequence = sequence + 1
    else
        local packet = encode({kind = "effects", requests = {{id = 1, kind = "compiler-request", result = result}}})
        phase("encode-end")
        local responseText = coroutine.yield(packet)
        sequence = sequence + 1
        phase("decode-start")
        local response = decode(responseText)
        request = assert(response.responses[1].value)
    end
    phase("decode-end")
    if request.stop then
        return {ok = true}
    end
    phase("execute-start")
    local started = now()
    local cpu = os.clock()
    local answers = {}
    for i = 1, request.count do
        if request.kind == "noop" then
            answers[i] = {bytes = #(request.padding or "")}
        elseif request.kind == "busy" then
            print("COMPILER_BUSY_BEGIN")
            io.flush()
            while true do
            end
        else
            local source = request.source or "local value: number = 1; return value"
            if request.kind == "edit" then
                source = source:gsub("= 1", "= " .. i)
            end
            local kind = request.kind == "edit" and "check" or request.kind
            if config.direct then
                if kind == "hover" then
                    answers[i] = session:hover(7)
                else
                    answers[i] = session[kind](session, source, "latency.g.nupp", {dialect = "luajit"})
                end
            else
                answers[i] = decode(
                    session:request(
                        encode({
                            kind = kind,
                            source = source,
                            filename = "latency.g.nupp",
                            offset = 7,
                            options = {dialect = "luajit"}
                        })
                    )
                )
            end
        end
    end
    phase("execute-end")
    result = {
        sampledWallMs = now() - started,
        guestCpuMs = (os.clock() - cpu) * 1000,
        count = request.count,
        answers = answers
    }
end
